File: client-clojure/src/sirqul_io_t_platform/specs/account.clj
(ns sirqul-io-t-platform.specs.account
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.region :refer :all]
            [sirqul-io-t-platform.specs.third-party-credentials :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.cell-carrier :refer :all]
            [sirqul-io-t-platform.specs.address :refer :all]
            [sirqul-io-t-platform.specs.account-profile-info :refer :all]
            [sirqul-io-t-platform.specs.retailer-location :refer :all]
            )
  (:import (java.io File)))


(def account-data
  {
   (ds/opt :id) int?
   (ds/opt :created) inst?
   (ds/opt :updated) inst?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   (ds/opt :locationDescription) string?
   (ds/opt :currentRegion) region-spec
   (ds/opt :locationLastUpdated) inst?
   (ds/opt :currentPostalCode) string?
   (ds/opt :currentPostalCodeLatitude) float?
   (ds/opt :currentPostalCodeLongitude) float?
   (ds/opt :locationUpdated) boolean?
   (ds/opt :premigrationId) int?
   (ds/opt :username) string?
   (ds/opt :locale) string?
   (ds/opt :inviteCode) string?
   (ds/opt :retailerInviteCode) string?
   (ds/opt :aboutUs) string?
   (ds/opt :businessName) string?
   (ds/opt :likeCount) int?
   (ds/opt :dislikeCount) int?
   (ds/opt :thirdPartyCredentials) (s/coll-of third-party-credentials-spec)
   (ds/opt :partnerActive) boolean?
   (ds/opt :resetToken) string?
   (ds/opt :profileImage) string?
   (ds/opt :accountType) string?
   (ds/opt :media) (s/coll-of asset-spec)
   (ds/opt :picture) asset-spec
   (ds/opt :firstName) string?
   (ds/opt :middleName) string?
   (ds/opt :lastName) string?
   (ds/opt :prefix) string?
   (ds/opt :suffix) string?
   (ds/opt :title) string?
   (ds/opt :homePhone) string?
   (ds/opt :cellPhone) string?
   (ds/opt :cellCarrier) cell-carrier-spec
   (ds/opt :businessPhone) string?
   (ds/opt :businessPhoneExt) string?
   (ds/opt :faxNumber) string?
   (ds/opt :timeZone) string?
   (ds/opt :utcOffset) string?
   (ds/opt :code501c3) string?
   (ds/opt :emailAddress) string?
   (ds/opt :address) address-spec
   (ds/opt :web) string?
   (ds/opt :display) string?
   (ds/opt :website) string?
   (ds/opt :consumersNotes) string?
   (ds/opt :consumersAboutUs) string?
   (ds/opt :accountProfile) account-profile-info-spec
   (ds/opt :retailerLocations) (s/coll-of retailer-location-spec)
   (ds/opt :executive) boolean?
   (ds/opt :contentName) string?
   (ds/opt :defaultThreshold) int?
   (ds/opt :audienceInfo_gender) string?
   (ds/opt :flagThreshold_id) int?
   (ds/opt :contentType) string?
   })

(def account-spec
  (ds/spec
    {:name ::account
     :spec account-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/Account.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Account._

case class Account (
  id: Option[Long],
created: Option[OffsetDateTime],
updated: Option[OffsetDateTime],
active: Option[Boolean],
valid: Option[Boolean],
latitude: Option[Double],
longitude: Option[Double],
locationDescription: Option[String],
currentRegion: Option[Region],
locationLastUpdated: Option[OffsetDateTime],
currentPostalCode: Option[String],
currentPostalCodeLatitude: Option[Double],
currentPostalCodeLongitude: Option[Double],
locationUpdated: Option[Boolean],
premigrationId: Option[Long],
username: Option[String],
locale: Option[String],
inviteCode: Option[String],
retailerInviteCode: Option[String],
aboutUs: Option[String],
businessName: Option[String],
likeCount: Option[Long],
dislikeCount: Option[Long],
thirdPartyCredentials: Option[List[ThirdPartyCredentials]],
partnerActive: Option[Boolean],
resetToken: Option[String],
profileImage: Option[String],
accountType: Option[String],
media: Option[List[Asset]],
picture: Option[Asset],
firstName: Option[String],
middleName: Option[String],
lastName: Option[String],
prefix: Option[String],
suffix: Option[String],
title: Option[String],
homePhone: Option[String],
cellPhone: Option[String],
cellCarrier: Option[CellCarrier],
businessPhone: Option[String],
businessPhoneExt: Option[String],
faxNumber: Option[String],
timeZone: Option[String],
utcOffset: Option[String],
code501c3: Option[String],
emailAddress: Option[String],
address: Option[Address],
web: Option[String],
display: Option[String],
website: Option[String],
consumersNotes: Option[String],
consumersAboutUs: Option[String],
accountProfile: Option[AccountProfileInfo],
retailerLocations: Option[List[RetailerLocation]],
executive: Option[Boolean],
contentName: Option[String],
defaultThreshold: Option[Long],
audienceInfoGender: Option[AudienceInfoGender],
flagThresholdId: Option[Long],
`contentType`: Option[String])

object Account {
  import DateTimeCodecs._
  sealed trait AudienceInfoGender
  case object MALE extends AudienceInfoGender
  case object FEMALE extends AudienceInfoGender
  case object ANY extends AudienceInfoGender

  object AudienceInfoGender {
    def toAudienceInfoGender(s: String): Option[AudienceInfoGender] = s match {
      case "MALE" => Some(MALE)
      case "FEMALE" => Some(FEMALE)
      case "ANY" => Some(ANY)
      case _ => None
    }

    def fromAudienceInfoGender(x: AudienceInfoGender): String = x match {
      case MALE => "MALE"
      case FEMALE => "FEMALE"
      case ANY => "ANY"
    }
  }

  implicit val AudienceInfoGenderEnumEncoder: EncodeJson[AudienceInfoGender] =
    EncodeJson[AudienceInfoGender](is => StringEncodeJson(AudienceInfoGender.fromAudienceInfoGender(is)))

  implicit val AudienceInfoGenderEnumDecoder: DecodeJson[AudienceInfoGender] =
    DecodeJson.optionDecoder[AudienceInfoGender](n => n.string.flatMap(jStr => AudienceInfoGender.toAudienceInfoGender(jStr)), "AudienceInfoGender failed to de-serialize")

  implicit val AccountCodecJson: CodecJson[Account] = CodecJson.derive[Account]
  implicit val AccountDecoder: EntityDecoder[Account] = jsonOf[Account]
  implicit val AccountEncoder: EntityEncoder[Account] = jsonEncoderOf[Account]
}
